### PR TITLE
Run checks

### PR DESCRIPTION
**Pull request by Topvennie**
DELETE YOUR `.env` FILE
DELETE YOUR DATABASE
RUN `./development -c`

### Route Changes

- Removed
	- POST /api/projects (you should use POST /api/courses/\<id\>/projects/
	- POST & PUT & PATCH & DELETE /api/submissions/\<id\>/
- Moved
	- PUT & PATCH /api/courses/\<id\>/projects to /api/projects/\<id\>/
	- PUT & PATCH /api/projects/\<id\>/structure-checks to /api/structure-checks/\<id\>/

### Submissions

A submission is zipped upon receiving. 
No matter the content received.
This can lead to a zip inside a zip but no problem!

### Zip Parsing 

- Supports blocked extensions. 
	- Precede the extension with a `-` to make it a blocked extension.
		- For example the file `test.-py` will result in the blocked extension `py`.

### Checks

Extra checks are run after structure checks and only if the structure checks succeed.
Teachers / assistans have the ability to trigger the desired error message by exiting their bash script with the right exit code.

### Type Checking

For the chad VSCode Pylance users, the amount of typing errors decreased drastically :tada:

### Models

- Checks
	- Moved FileExtensions to `checks.py`
	
- Submission
	- Added the zip field, Replaces the `SubmissionFile` model.

- Check Result
	- Added
		- FILE_DIR_NOT_FOUND
			- An obligated file or directory during the structure check wasn't found
		- CHECK_ERROR
			- Can only be thrown by a teacher / assistant during an extra check
			- Should be used to signal that a custom check failed (equivalent of the Dodona checks)
		- FAILED_STRUCTURE_CHECK
			- Error message used by `ExtraCheckResult` to indicate that it didn't run because the structure checks didn't succeed.
		- DOCKER_IMAGE_ERROR
			- The docker image that was selected to use for the extra check isn't ready yet (should never be possible).
	- Removed
		- OBLIGATED_DIRECTORY_NOT_FOUND
		- UNASKED_DIRECTORY
		- OUTPUTLIMIT
		- INTERNALERROR
	- Renamed
		- TIMELIMIT -> TIME_LIMIT
		- MEMORYLIMIT -> MEMORY_LIMIT
		- CHECKERROR -> CHECK_ERROR
		- RUNTIMEERROR -> RUNTIME_ERROR
		
- StructureCheck
	- Renamed
		- Name -> Path

- Docker Image
	- Added the field `state`. Can be one of the following:
		- Queued
		- Building
		- Ready
		- Error
		
- Extra Check
	- Added
		- Name
		- Memory_limit
			- Maximum amount of ram a docker container can use in MB.
			- Default is 128 MB. Can be increased to 1024 MB.
			- Should not be used by a teacher / assistant to limit the amount of ram a submission can use as it will force stop the entire container.
	- Renamed
		- Timeout -> time_limit
			- Maximum time a docker container can run in seconds.
			- Default is 30 seconds. Can be increased to 1000.
			- Goal is to stop infinite loops as quickly as possible.
			- Should not be used by a teacher / assistant to time a submission as it will force stop the entire container.

### CODE REVIEW

File: backend/api/serializers/project_serializer.py [hunk 1]
```diff
@@ -102,67 +100,63 @@ def validate(self, data):
...
-class CreateProjectSerializer(ProjectSerializer):
-    number_groups = serializers.IntegerField(min_value=0, required=False)
-    zip_structure = serializers.FileField(required=False, read_only=True)
-
     def create(self, validated_data):
         # Pop the 'number_groups' field from validated_data
         number_groups = validated_data.pop('number_groups', None)
 
         # Get the zip structure file from the request
+        # TODO: ?
```
Comment (DeLany123) [inline backend/api/serializers/project_serializer.py:113]: ?

Comment (Topvennie) [reply backend/api/serializers/project_serializer.py:113]: Will be removed in a future commit.

File: backend/api/views/project_view.py [hunk 2]
```diff
@@ -100,30 +99,27 @@ def structure_checks(self, request, **_):
...
     @structure_checks.mapping.post
-    @structure_checks.mapping.put
     @swagger_auto_schema(request_body=StructureCheckAddSerializer)
     def _add_structure_check(self, request: Request, **_):
         """Add a structure_check to the project"""
 
         project: Project = self.get_object()
 
-        # TODO: Crashes when requesting without any fields, shouldn't be given as context. Validator needs to check it
+        print(request.data, flush=True)
```
Comment (DeLany123) [inline backend/api/views/project_view.py:108]: !

**Review by DeLany123 — CHANGES_REQUESTED** [review]
Because of the change to the structureCheckModel, the fixtures do not work anymore (and seeder probably too). It would be nice if these worked first so the frontend developers do not experience errors.

**Comment by DeLany123** [comment]
Could it be possible to make a working example? 
By this I mean making it possible to to add a submission where there are working extra checks on the background. This should include a working docker-image, working dockerfile (check from the teacher) and a working submissionfile which we can use to test a submission. In 5f176ab I've tried to make this work myself where you can find the files that I tried making under backend/api/fixtures/data.

**Comment by DeLany123** [comment]
Changes requested here!

**Comment by Topvennie** [comment]
> Changes requested here!

:+1:

**Comment by DeLany123** [comment]
Very good very nice!

**Review by DeLany123 — APPROVED** [review]

**Pull request merged**